Blockout/fullscreen guis

Nightenom opened the pull request:
Creates #238
Closes #238
Closes #243
Closes #244

# Changes proposed in this pull request:
- added a render type to Window XML element, default is VANILLA
- VANILLA is 320*240 px
- FIXED is always gui sized
- OVERSIZED is integer upscaling like vanilla gui size setting or downscaling to framebuffer size
- FULLSCREEN adapts to framebuffer size so it matches current window size, distortion prevented
- _VANILLA suffix blocks downscaling under 320*240px 
- fixed scissors in multi scaled environment (switched to matrixstack calculations), fixed vanilla buttons not rendering correctly for huge sizes
- fixed ZoomDragView not correctly reading pane params, and not correctly scaling some mouse methods
- added textscale, textalign and shadow to vanilla buttons, they now wrap label if needed
- added blitRepeatable method for rendering of texture with repeatable interior
- changed mouse pos args in draw methods from int to double
- @Raycoms use this ˇˇ for the broken gui in minecolonies, it should be prepared
```
<window xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../../../../src/main/java/com/ldtteam/blockout/blockOut.xsd"
    pause="false" size="960 540" type="FULLSCREEN_VANILLA" lightbox="false" >

    <image source="structurize:textures/gui/background.png" size="960 540" id="guide1"/>
    <button id="confirm" pos="360 455" label="Confirm"/>
    <text id="questionmark" size="210 75" color="black" pos="80 363" textscale="0.8"/>
    <text id="arrow" size="210 75" color="black" pos="637 22" textscale="0.8"/>
    <text id="chest" size="210 75" color="black" pos="646 363" textscale="0.8"/>
</window>
```

Review please

EDIT: edited to match current PR changes

Nightenom committed:
Double the mouse, double the cheese loss

Nightenom committed:
Add fullscreen gui mode

scissors need fix

Nightenom committed:
Update rendering pipeline with recent vanilla changes

Nightenom committed:
New scissors, good steel, sharp as shark

Nightenom commented:
I'm just dumb, we can use matrix stack now for easy scissor calculation

Nightenom committed:
Fix vanilla rendering, added scissors boxes to debug rendering

Nightenom committed:
Code cleanup

Nightenom committed:
Update BOScreen.java

Nightenom committed:
Add oversized, another code cleanup

Nightenom committed:
Fix mistake

marchermans commented on src/main/java/com/ldtteam/blockout/BOScreen.java:
```diff
@@ -111,44 +114,49 @@ public BOScreen(final Window w)
-        ms.pop();
+        final float renderZlevel = MatrixUtils.getLastMatrixTranslateZ(ms);
+        final float oldZ = minecraft.getItemRenderer().zLevel;
+        minecraft.getItemRenderer().zLevel = renderZlevel;
```
You need a none null check for minecraft in all this code.

Nightenom replied:
Then `if minecraft == null return` and we do no rendering

marchermans replied:
You could add that too.

marchermans commented on src/main/java/com/ldtteam/blockout/BOScreen.java:
```diff
@@ -111,44 +114,49 @@ public BOScreen(final Window w)
+        x = Math.floor((guiWidth - window.getWidth() * renderScale) / 2.0d);
+        y = Math.floor((guiHeight - window.getHeight() * renderScale) / 2.0d);
+
+        RenderSystem.matrixMode(GL11.GL_PROJECTION);
```
Not sure this is needed here.

Nightenom replied:
Hmm?

marchermans replied:
It at least requires some documentation.

Nightenom replied:
Literally basic opengl rendering setup, but ok

marchermans replied:
No why is it here.
Mc already applies this when rendering uis
So why do you need to reset everything again?
That is what I am wondering

Nightenom replied:
It's not resetting, you just replace projection with unscaled one and do the scaling in modelview stack, you can probably set this up also in modelview stack but this is way safer

Nightenom replied:
Like you can do in modelview stack
ms.getLast.getMatrix.loadIdentity
ms.translate x y zLevel
ms.scale renderScale/mcScale
but you would need to fix scissors since they would still be in mcScale environment rather than in framebuffer scaled environment

Nightenom committed:
Update BOScreen.java

Raycoms committed:
add pass to this handler

Nightenom committed:
Fix vanilla button rendering when size is huge

Now renders vanilla button using repeating inside of vanilla texture rather than scaling

Nightenom committed:
Fix vanilla button rendering, added textScale and textAlign to vanilla button

Nightenom committed:
Update ButtonVanilla.java

Nightenom committed:
Update ButtonVanilla.java

Nightenom committed:
fix label offset

Nightenom committed:
add tons of window types

Nightenom committed:
Update ItemCaliper.java

Raycoms committed:
fix fluid placement

Raycoms committed:
Merge remote-tracking branch 'origin/version/1.16.3' into blockout/fullscreen-guis

Raycoms committed:
water

Raycoms committed:
oopsy

Raycoms commented on src/main/java/com/ldtteam/blockout/Pane.java:
```diff
@@ -856,4 +876,66 @@ public void setHoverToolTip(final List<IFormattableTextComponent> lines)
+        final int uWidth, final int vHeight, // texture rendering box [texels]
+        final int textureWidth, final int textureHeight, // texture file size [texels]
+        final int uRepeat, final int vRepeat, // offset relative to u, v [texels]
+        final int repeatWidth, final int repeatHeight) // size of repeatable box in texture [texels]
```
this param doc should be javadoc @param

Raycoms commented on src/main/java/com/ldtteam/structurize/items/ItemCaliper.java:
```diff
@@ -45,6 +45,8 @@ public ActionResultType onAirRightClick(final BlockPos start,
         }
+        // fullscreen gui test
+        // new com.ldtteam.structurize.client.gui.WindowFullscreenTest().open(); if(true)return ActionResultType.SUCCESS;
 
```
remove comment

Nightenom committed:
Remove comment, spam log if excludedEntities config is used

Nightenom committed:
Add comment to config, modify default values and don't shout on old default values

Nightenom committed:
Delete test.xml

Nightenom committed:
Delete WindowFullscreenTest.java

Nightenom committed:
Merge branch 'version/1.16.3' into blockout/fullscreen-guis

Nightenom committed:
Try to fix incorrect hoverText positioning

Nightenom committed:
Remove obsolete hover mechanism, fixed scroll not correctly adjusting mouse pos

No need to perform dfs every mouse move

Nightenom committed:
Few fixes

Nightenom committed:
Add generic repeatable image, buffer blitRepeatable into one buffer

Nightenom committed:
Bunch of bug fixes

Nightenom committed:
Reverts and code cleanup

Nightenom committed:
Fix label autosetup

ShadowProtocol commented on src/main/java/com/ldtteam/blockout/views/View.java:
```diff
@@ -94,12 +100,12 @@ public void drawSelfLast(final MatrixStack ms, final int mx, final int my)
         for (final Pane child : children)
         {
-            if (child.isVisible())
+            if (childIsVisible(child) && child.isVisible())
```
A very small thing, but I had to look twice to notice that they were not the exact same condition. It might be helpful to rename the method to ``childWithinBounds`` or something?

Nightenom replied:
this mechanism probably needs a visit in the future, so can be renamed later, gonna create issue

ShadowProtocol commented on src/main/java/com/ldtteam/blockout/views/ZoomDragView.java:
```diff
@@ -42,12 +43,12 @@ public ZoomDragView()
-        zoomEnabled = params.getBooleanAttribute("zoomenabled", zoomEnabled);
-        minScale = params.getDoubleAttribute("minscale", minScale);
-        maxScale = params.getDoubleAttribute("maxscale", maxScale);
+        dragFactor = params.getDoubleAttribute("dragFactor", dragFactor);
```
Maybe (probably) its just me, but I expect all markup attributes to be in lowercase or kebab-case. To me this is less guessable. I know the standard does have them, but I haven't really seen it round too often, bar the schema one.

Raycoms replied:
yeah, the others are maxscale so this should be dragfactor

Nightenom replied:
kk, gonna change the schema

ShadowProtocol commented on src/main/resources/assets/structurize/lang/en_us.json:
```diff
@@ -1151,5 +1151,7 @@
   "structurize.gui.buildtool.creative_only": "Structurize does not support using the build tool when in survival. Switch to creative or install MineColonies and use the MineColonies Builder.",
-  "structurize.gui.replaceblock.ambiguous_properties": "Transformation from %s to %s is ambiguous because the following properties are not present in the source block: %s"
+  "structurize.gui.replaceblock.ambiguous_properties": "Transformation from %s to %s is ambiguous because the following properties are not present in the source block: %s",
+  "structurize.config.excludeentities": "Entities excluded from BP rendering",
```
What's BP? Just so every user knows explicitly.

ravenbuilder934 replied:
I agree, blueprint should be spelled out

```suggestion
  "structurize.config.excludeentities": "Entities Excluded from Blueprint Rendering",
```

Nightenom replied:
I agree with this but the name is just long, but could not find something shorter
Feel free to suggest something completely different

Raycoms replied:
It's in the config, doesn't matter if its a bit longer

Nightenom replied:
config name is gonna be used for gui buttons, so they can't be long

ShadowProtocol reviewed (commented):
Haven't had a good study into the scissor stuff, but this looks good to me.

ravenbuilder934 commented on src/main/resources/assets/structurize/lang/en_us.json:
```diff
@@ -1151,5 +1151,7 @@
-  "structurize.gui.replaceblock.ambiguous_properties": "Transformation from %s to %s is ambiguous because the following properties are not present in the source block: %s"
+  "structurize.gui.replaceblock.ambiguous_properties": "Transformation from %s to %s is ambiguous because the following properties are not present in the source block: %s",
+  "structurize.config.excludeentities": "Entities excluded from BP rendering",
+  "structurize.config.excludeentities.comment": "Excludes entities from blueprint rendering if they can't load on client side. If you are going to add \"minecraft\" entity or you believe it's our fault that an entity couldn't be set up, please, make an issue at our GitHub and add stacktrace to it (so we can try to fix it)."
```
```suggestion
  "structurize.config.excludeentities.comment": "Excludes entities from blueprint rendering if they can't load clientside. If you are going to add a "\minecraft\" entity or you believe it's our fault that an entity couldn't be set up, please make an issue at our GitHub and add the stacktrace to it."
```

ShadowProtocol commented on src/main/java/com/ldtteam/blockout/controls/ImageRepeatable.java:
```diff
@@ -0,0 +1,155 @@
+/**
+ * Image element with repeatable middle part.
+ */
+public class ImageRepeatable extends Pane
```
Surely we can extend the Image functionality, if not integrate the two?

Nightenom replied:
image is scalable, this is not (or not supposed to be), so integrating can be difficult, but for sure something can be abstracted
same goes towards ButtonImage - there is a problem of having the image logic three times duplicated
leaving the decision up to you

ShadowProtocol replied:
Hmm, I think my PR will solve this anyway, it will allow for property grouping of sorts. I'll let it explain itself once I push it.

Nightenom replied:
Marking as resolved for now

ShadowProtocol commented on src/main/java/com/ldtteam/blockout/controls/ImageRepeatable.java:
```diff
@@ -0,0 +1,155 @@
+            loadMapDimensions();
+        }
+
+        PaneParams.SizePair size = params.getSizePairAttribute("uv", null, null);
```
We should probably stick with the "imageoffset" and "imagesize" terminology for consistency. Personally I'd like to use the terms "offset" and "crop" when referring to the texture throughout all of blockout, but that's just my personal sense of clarity.

Nightenom replied:
I would personally rename the old params, since imageOffset is more like pane offset for me
crop means something different? or I haven't seen it being used in this context

ShadowProtocol replied:
I have nothing wrong with offset personally. Could just name everything with a source prefix, i.e. ``sourcepos`` ``sourcesize``, which is the most boring but most clear for newcomers I think. Technically we are cropping out a section of the source, but you're right in that crop doesn't usually refer directly to the width and height. I have seen it used here before though, if only once. We could potentially code in a 4-hand property, e.g. ``crop="2 2 192 20"``, which is a more logical use of the word (but of course it's ambiguous whether it is point-to-point or pos and size).

I don't really mind which but I do think we should be consistent across the board.

Nightenom replied:
> We could potentially code in a 4-hand property, e.g. crop="2 2 192 20", which is a more logical use of the word (but of course it's ambiguous whether it is point-to-point or pos and size).

I have thought about this too, not sure if it's a good approach tho

We can also go back to traditional `textureoffset` and `texturesize`, and thus `repeatoffset `and `repeatsize`

ShadowProtocol replied:
Yeah that's nice and clear, I like it.

ShadowProtocol commented on src/main/java/com/ldtteam/blockout/controls/ImageRepeatable.java:
```diff
@@ -0,0 +1,155 @@
+            vHeight = size.getY();
+        }
+
+        size = params.getSizePairAttribute("uvRepeat", null, null);
```
just a non-critical point of discussion: could we use "slice" instead of defining a box in a box. The slice will define the number of pixels in from each edge to go to get your nine-square layout. Personally I find it more intuitive, but maybe that's because it's how I first learned tiling systems.

Nightenom replied:
box in box is more customizable, your slice has fixed all borders to the same size, if I get you right

ShadowProtocol replied:
You could have it as a pair for vertical and horizontal, or as a 4-hand property for each edge. Still my preference but I don't mind either way. I would note though that uvRepeat isn't immediately apparent of what it refers to, though I'm not sure how to make that much clearer.

ShadowProtocol reviewed (commented):
Just a couple of opinionated things from me, but this looks good. 
Hopefully some of my changes can make things a bit DRYer (not that they're that bad).

Nightenom committed:
Blockout/rework text elements (#247) and Gui hooks for entities and tileentities (#249) and Tooltip element (#250)

Nightenom committed:
Review

Nightenom committed:
Merge branch 'blockout/fullscreen-guis' of https://github.com/ldtteam/Structurize into blockout/fullscreen-guis

Nightenom committed:
Merge branch 'version/1.16.3' into blockout/fullscreen-guis

Nightenom commented:
This PR is now in frozen state so it does not prevent fixes in mcol. Waiting for Shadow to do his changes and then this will get merged upstream

ravenbuilder934 committed:
Merge branch 'version/1.16.3' into blockout/fullscreen-guis

ShadowProtocol committed:
Blockout properties (#258)

Nightenom committed:
Fixes and cleanup
